File: haskell/twelve-days/src/TwelveDays.hs
module TwelveDays (recite) where

-- join a list of items together using "," and "and"
-- eg join "dog" "cat" "mouse" => "dog, cat, and mouse"
join :: [String] -> String
join xs =
  foldl (\x y -> x ++ y ++ ", " ) "" (init xs)
  ++ "and " ++ last xs

-- dayItems gives the list of items for a given day.
dayItems :: Int -> String
dayItems n
  | n > 1 = join . reverse . take n $ items
  | otherwise = head items
  where
    items = [
      "a Partridge in a Pear Tree", "two Turtle Doves",
      "three French Hens", "four Calling Birds",
      "five Gold Rings", "six Geese-a-Laying",
      "seven Swans-a-Swimming", "eight Maids-a-Milking",
      "nine Ladies Dancing", "ten Lords-a-Leaping",
      "eleven Pipers Piping", "twelve Drummers Drumming"]

-- recite the verse for a given day.
reciteDay :: Int -> String
reciteDay n =
  "On the "
  ++ (position !! (n - 1))
  ++ " day of Christmas my true love gave to me: "
  ++ dayItems n
  ++ "."
  where
    position = ["first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"]

recite :: Int -> Int -> [String]
recite start stop = map reciteDay [start .. stop]


-- vim:ts=2:sw=2:expandtab

File: python/twelve-days/twelve_days.py
#!/bin/python

opening = 'On the %s day of Christmas my true love gave to me: '
item = [
  'zero',
  'and a Partridge in a Pear Tree',
  'two Turtle Doves',
  'three French Hens',
  'four Calling Birds',
  'five Gold Rings',
  'six Geese-a-Laying',
  'seven Swans-a-Swimming',
  'eight Maids-a-Milking',
  'nine Ladies Dancing',
  'ten Lords-a-Leaping',
  'eleven Pipers Piping',
  'twelve Drummers Drumming',
]
numbers = [
  'zeroth', 'first', 'second', 'third', 'fourth', 'fifth', 'sixth',
  'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth']


def verse(i):
  out = opening % numbers[i]
  if i == 1:
    items = ['a Partridge in a Pear Tree']
  else:
    items = item[i:0:-1]
  out += ', '.join(items) + '.'
  return out

  
def recite(start, end):
    return [verse(i) for i in range(start, end + 1)]


# vim:ts=2:sw=2:expandtab
